clang/lib/Sema/SemaTemplateInstantiate.cpp

```diff
@@ -3437,10 +3437,10 @@ bool Sema::SubstDefaultArgument(
       //   template<typename T> void f(T a, int = decltype(a)());
       //   void g() { f(0); }
       LIS = std::make_unique<LocalInstantiationScope>(*this);
-      FunctionDecl *PatternFD = FD->getTemplateInstantiationPattern(
-          /*ForDefinition*/ false);
-      if (addInstantiatedParametersToScope(FD, PatternFD, *LIS, TemplateArgs))
-        return true;
+      if (FunctionDecl *PatternFD =
+              FD->getTemplateInstantiationPattern(/*ForDefinition*/ false))
+        if (addInstantiatedParametersToScope(FD, PatternFD, *LIS, TemplateArgs))
+          return true;
```

Comment:
How about lifting the check for PatternFD to ForCallExpr?

```cpp
if (FunctionDecl *PatternFD = FD->getTemplateInstantiationPattern(); ForCallExpr && ForCallExpr)
  ...
```

So we don't need to construct an extra LocalInstantiationScope if the pattern isn't present.

Reply:
OTOH I'm wondering if we could avoid substituting into default arguments if we already know the friend declaration is not a definition. The function declaration instantiation seems to intentionally avoid adding connections to the primary template if the friend declaration is not a definition, which is why `getTemplateInstantiationPattern()` returns nullptr here.

Reply:
@zyn0217 thanks for the feedback. 

> How about lifting the check for PatternFD to ForCallExpr?

A check for `ForCallExpr` is already present above
https://github.com/llvm/llvm-project/blob/1e19f0f9d92b5e9c43d53893e387341835d3d96b/clang/lib/Sema/SemaTemplateInstantiate.cpp#L3433

Reply:
I mean, if we could move the pointer check to the same line as ForCallExpr, we could save an LIS then.

Reply:
@zyn0217 whenever you get a chance, could you take a look at the latest changes? thanks